tm2/pkg/bft/state/txindex/indexer_service.go

```diff
@@ -10,22 +14,73 @@ import (
 type IndexerService struct {
 	service.BaseService
 
-	idr  TxIndexer
-	evsw events.EventSwitch
+	cancelFn context.CancelFunc
+
+	indexer  TxIndexer
+	evSwitch events.EventSwitch
 }
 
 // NewIndexerService returns a new service instance.
-func NewIndexerService(idr TxIndexer, evsw events.EventSwitch) *IndexerService {
-	is := &IndexerService{idr: idr, evsw: evsw}
+func NewIndexerService(idr TxIndexer, evSwitch events.EventSwitch) *IndexerService {
+	is := &IndexerService{indexer: idr, evSwitch: evSwitch}
 	is.BaseService = *service.NewBaseService(nil, "IndexerService", is)
+
 	return is
 }
 
 func (is *IndexerService) OnStart() error {
-	// TODO
+	// Create a context for the intermediary monitor service
+	ctx, cancelFn := context.WithCancel(context.Background())
+	is.cancelFn = cancelFn
+
+	// Start the indexer
+	if err := is.indexer.Start(); err != nil {
+		return fmt.Errorf("unable to start transaction indexer, %w", err)
+	}
+
+	// Start the intermediary monitor service
+	go is.monitorTxEvents(ctx)
+
 	return nil
 }
 
 func (is *IndexerService) OnStop() {
-	// TODO
+	// Close off any routines
+	is.cancelFn()
+
+	// Attempt to gracefully stop the transaction indexer
+	if err := is.indexer.Stop(); err != nil {
+		is.Logger.Error(
+			fmt.Sprintf("unable to gracefully stop transaction indexer, %v", err),
+		)
+	}
+}
+
+// monitorTxEvents acts as an intermediary feed service for the supplied
+// transaction indexer. It relays transaction events that come from the event stream
+func (is *IndexerService) monitorTxEvents(ctx context.Context) {
+	// Create a subscription for transaction events
+	subCh := events.SubscribeToEvent(is.evSwitch, "tx-indexer", types.EventTx{})
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case evRaw := <-subCh:
+			// Cast the event
+			ev, ok := evRaw.(types.EventTx)
+			if !ok {
+				is.Logger.Error("invalid transaction result type cast")
+
+				continue
+			}
+
+			// Alert the actual indexer
+			if err := is.indexer.Index(ev.Result); err != nil {
+				is.Logger.Error(
+					fmt.Sprintf("unable to index transaction, %v", err),
+				)
```

Comment:
According to the method signature, I think it's possible to write it like that : 

```suggestion
				is.Logger.Error("unable to index transaction, %v", err)
```

Reply:
Rewrote the log line to follow codebase convention:
[86c11c9](https://github.com/gnolang/gno/pull/546/commits/86c11c9f855f8fde945374e19a357f48dd06a041)

Reply:
oh right, I didn't see it was key-value pairs.